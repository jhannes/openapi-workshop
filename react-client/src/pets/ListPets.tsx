import * as React from "react";
import { useContext, useState } from "react";
import { PetDtoStatusEnum } from "../generated";
import { ApiContext } from "../applicationContext";
import { useLoader } from "../lib/useLoader";
import { Link } from "react-router-dom";
import { CheckboxList } from "../views/CheckboxList";
import { LoadingView } from "../views/LoadingView";
import { ErrorView } from "../views/ErrorView";
import { useApplicationTexts } from "../localization";

function ShowPetList({ statuses }: { statuses: PetDtoStatusEnum[] }) {
  const {
    apis: { petApi },
  } = useContext(ApiContext);

  const pets = useLoader(
    () => petApi.findPetsByStatus({ queryParams: { status: statuses } }),
    [statuses]
  );

  if (pets.loading) {
    return <LoadingView />;
  }
  if (pets.failed) {
    return <ErrorView error={pets.error} />;
  }

  return (
    <div>
      {pets.data.map(({ id, name }) => (
        <div key={id}>{name}</div>
      ))}
    </div>
  );
}

export function ListPets() {
  const [statuses, setStatuses] = useState<PetDtoStatusEnum[]>([]);

  const { petstoreTexts: texts } = useApplicationTexts();

  return (
    <div>
      <h1>{texts.showPets}</h1>
      <Link to={"/pets/new"}>
        <button>{texts.createPet}</button>
      </Link>
      <CheckboxList values={texts.statuses} setValues={setStatuses} />
      <ShowPetList statuses={statuses} />
    </div>
  );
}
