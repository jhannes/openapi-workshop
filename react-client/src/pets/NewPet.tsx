import * as React from "react";
import { useContext, useState } from "react";
import { ApiContext } from "../applicationContext";
import { useHistory } from "react-router";
import {
  PetDtoStatusDtoEnum,
  PetDtoStatusDtoEnumValues,
} from "@jhannes/openapi-workshop";
import { useLoader } from "../lib/useLoader";
import { useSubmit } from "../lib/useSubmit";
import { LoadingView } from "../views/LoadingView";
import { ErrorView } from "../views/ErrorView";
import { FormInputField } from "../views/FormInputField";
import { FormField } from "../views/FormField";

export function NewPet() {
  const {
    apis: { petApi },
    security,
  } = useContext(ApiContext);
  const history = useHistory();

  const [status, setStatus] = useState<PetDtoStatusDtoEnum>("available");
  const [categoryId, setCategoryId] = useState("");
  const [name, setName] = useState("");
  const [tags, setTags] = useState<string[]>([]);

  const categories = useLoader(async () => {
    return await petApi.listCategories();
  });

  const { handleSubmit, submitting, submitError } = useSubmit(
    async () => {
      return await petApi.addPet({
        petDto: { status, category: { id: categoryId }, name, tags },
        security,
      });
    },
    () => {
      history.push("/pets");
    }
  );

  if (categories.loading) {
    return <LoadingView />;
  }
  if (categories.failed) {
    return <ErrorView error={categories.error} />;
  }

  return (
    <div>
      <h1>New pet</h1>
      <form onSubmit={handleSubmit}>
        <fieldset disabled={submitting}>
          <FormInputField label={"Name"} value={name} onChangeValue={setName} />
          <FormField label={"Status"}>
            <select
              value={status}
              onChange={(e) => setStatus(e.target.value as PetDtoStatusDtoEnum)}
            >
              {PetDtoStatusDtoEnumValues.map((value) => (
                <option key={value} value={value}>
                  {value}
                </option>
              ))}
            </select>
          </FormField>
          <FormField label={"Category"}>
            <select
              value={categoryId}
              onChange={(e) => setCategoryId(e.target.value)}
            >
              <option value={""} />
              {categories.data.map(({ id, name }) => (
                <option key={id} value={id}>
                  {name}
                </option>
              ))}
            </select>
          </FormField>
          {submitError && <div>Error: {submitError.toString()}</div>}
          <button>Submit</button>
        </fieldset>
      </form>
    </div>
  );
}
