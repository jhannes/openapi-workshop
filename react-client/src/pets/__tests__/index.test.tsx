import { act, Simulate } from "react-dom/test-utils";
import { ListPets } from "../ListPets";
import * as React from "react";
import { activeDirectory } from "@jhannes/openapi-workshop";
import { mockPetApi } from "@jhannes/openapi-workshop/dist/test/apiTest";
import { NewPet } from "../NewPet";
import {
  renderApplication,
  simulateCheck,
  simulateFormChange,
  testData,
} from "../../__mocks__/applicationTest";
import pretty = require("pretty");

describe("pets screens", () => {
  it("shows new pet screen", async () => {
    const sampleData = testData(100);
    const container = await renderApplication(<NewPet />, {
      petApi: mockPetApi({
        listCategories: async () => sampleData.sampleArrayCategoryDto(),
      }),
    });
    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  it("submits new pet", async () => {
    const sampleData = testData(200);
    const categories = sampleData.sampleArrayCategoryDto();
    const petDto = {
      name: sampleData.samplePetDto().name,
      category: { id: categories[1].id },
      status: sampleData.samplePetDto().status,
    };
    const addPet = jest.fn();
    const security = new activeDirectory(sampleData.uuidv4());
    const container = await renderApplication(
      <NewPet />,
      {
        petApi: mockPetApi({
          addPet,
          listCategories: async () => categories,
        }),
      },
      security
    );
    simulateFormChange(container, "Name", petDto.name);
    simulateFormChange(container, "Category", petDto.category.id);
    simulateFormChange(container, "Status", petDto.status);
    await act(async () => {
      Simulate.submit(container.querySelector("form")!);
    });
    expect(addPet).toHaveBeenCalledWith({ petDto, security });
  });

  it("shows pet list", async () => {
    const sampleData = testData(200);
    const container = await renderApplication(<ListPets />, {
      petApi: mockPetApi({
        findPetsByStatus: async () => sampleData.sampleArrayPetDto(),
      }),
    });
    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  it("filters pet list", async () => {
    const sampleData = testData(200);
    const findPetsByStatus = jest.fn(async () =>
      sampleData.sampleArrayPetDto()
    );
    const container = await renderApplication(<ListPets />, {
      petApi: mockPetApi({ findPetsByStatus }),
    });
    await act(async () => {
      simulateCheck(container, "available", true);
    });
    expect(findPetsByStatus).toHaveBeenCalledWith({
      queryParams: { status: ["available"] },
    });
  });

  it("shows errors on new pet screen", async () => {
    const container = await renderApplication(<NewPet />, {
      petApi: mockPetApi({
        listCategories: async () => {
          throw new Error("Something went wrong");
        },
      }),
    });
    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  it("shows errors on pet list", async () => {
    const container = await renderApplication(<ListPets />, {
      petApi: mockPetApi({
        findPetsByStatus: async () => {
          throw new Error("Something went wrong");
        },
      }),
    });
    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });
});
