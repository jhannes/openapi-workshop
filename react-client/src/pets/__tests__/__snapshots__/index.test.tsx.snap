// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pets screens shows errors on new pet screen 1`] = `
"<div>
  <h1>An error has occurred</h1>
  <div>Error: Something went wrong</div>
</div>"
`;

exports[`pets screens shows errors on pet list 1`] = `
"<div>
  <h1>Show pets</h1><a href=\\"/pets/new\\"><button>Create pet</button></a>
  <div><label data-label=\\"available\\"><input type=\\"checkbox\\" class=\\"type\\">Available</label><label data-label=\\"pending\\"><input type=\\"checkbox\\" class=\\"type\\">Pending</label><label data-label=\\"sold\\"><input type=\\"checkbox\\" class=\\"type\\">Sold</label></div>
  <div>
    <h1>An error has occurred</h1>
    <div>Error: Something went wrong</div>
  </div>
</div>"
`;

exports[`pets screens shows new pet screen 1`] = `
"<div>
  <h1>New pet</h1>
  <form>
    <fieldset>
      <div><label data-label=\\"Name\\">Name: <input type=\\"text\\" value=\\"\\"></label></div>
      <div><label data-label=\\"Status\\">Status: <select>
            <option value=\\"available\\">available</option>
            <option value=\\"pending\\">pending</option>
            <option value=\\"sold\\">sold</option>
          </select></label></div>
      <div><label data-label=\\"Category\\">Category: <select>
            <option value=\\"\\"></option>
            <option value=\\"f5d54bf1-673b-410f-8665-8e1485992f4a\\">Cat</option>
          </select></label></div><button>Submit</button>
    </fieldset>
  </form>
</div>"
`;

exports[`pets screens shows pet list 1`] = `
"<div>
  <h1>Show pets</h1><a href=\\"/pets/new\\"><button>Create pet</button></a>
  <div><label data-label=\\"available\\"><input type=\\"checkbox\\" class=\\"type\\">Available</label><label data-label=\\"pending\\"><input type=\\"checkbox\\" class=\\"type\\">Pending</label><label data-label=\\"sold\\"><input type=\\"checkbox\\" class=\\"type\\">Sold</label></div>
  <div>
    <div>Chase</div>
    <div>Rocky</div>
  </div>
</div>"
`;
